runtime/include/map.h

```diff
@@ -0,0 +1,183 @@
+#pragma once
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "lock.h"
+
+/* Simple K-V store based on The Practice of Programming by Kernighan and Pike */
+
+/* Bucket count is sized to be a prime that is approximately 20% larger than the desired capacity (6k keys) */
+#define MAP_BUCKET_COUNT 7907
+#define MAP_LOCK_STRIPES 17
+
+#define MAP_HASH jenkins_hash
+
+struct map_node {
+	uint32_t         hash;
+	uint8_t         *key;
+	uint32_t         key_len;
+	uint8_t         *value;
+	uint32_t         value_len;
+	struct map_node *next;
+};
+
+struct map {
+	struct map_node *buckets[MAP_BUCKET_COUNT];
```

Comment:
A downside of this structure is that we're going to have a cache-line bounce for the lock, and a *separate* cache line bounce for the budget. As MCS locks are small (only a word?), inlining it into the `map_node` wouldn't take too much memory. Not sure where the trade-offs pay off here. I dont' think that we're as concerned about memory as java's maps as we have many fewer of these than they do.

Reply:
I added a new struct `map_bucket` that contains a lock and a pointer to the head node. My main concern was the overall complexity of adding a lock on each node. That seems to suggest hand-over-hand locking to me, which seems like it'd be overly expensive assuming collisions are rare.

Reply:
I added a new struct `map_bucket` that contains a lock and a pointer to the head node. My main concern was the overall complexity of adding a lock on each node. That seems to suggest hand-over-hand locking to me, which seems like it'd be overly expensive assuming collisions are rare.